File: problems/2022_09_27/0090/Solution.java
// NU

class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        backtrack(0, res, list, nums);
        return res;
    }

    private void backtrack(int index, List<List<Integer>> res, List<Integer> list, int[] nums) {
        res.add(new ArrayList<>(list));

        for (int i = index; i < nums.length; ++i) {
            if (i > index && nums[i] == nums[i - 1]) {
                continue;
            }
            list.add(nums[i]);
            backtrack(i + 1, res, list, nums);
            list.remove(list.size() - 1);
        }
    }
}

File: problems/2022_09_27/0090/Solution.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        self.res = []
        self.backtracking(nums, [], 0)

        return self.res

    def backtracking(self, nums, curr, i):
        if i == len(nums) and curr not in self.res:
            self.res.append(curr.copy())
            return

        if i >= len(nums):
            return

        curr.append(nums[i])
        self.backtracking(nums, curr, i + 1)

        curr.pop()
        while i + 1 < len(nums) and nums[i] == nums[i + 1]:
            i += 1
        self.backtracking(nums, curr, i + 1)
